### PR TITLE
Vb 4066 backend validations public booking mvp booking

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Adds validations during the booking process.

## What is the intent behind these changes?

VB-4066

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 1]
```diff
@@ -100,7 +97,9 @@ class VisitService(
     val application = applicationService.getApplicationEntity(applicationReference)
 
     val existingBooking = visitRepository.findVisitByApplicationReference(application.reference)
-    checkSlotCapacity(bookingRequestDto, application, existingBooking)
+
+    // application validity checks
+    applicationValidationService.isApplicationValid(bookingRequestDto, application, existingBooking)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:102]: We also need a call from the application service right?

**Review by AledWEvans — COMMENTED** [review]
have left some comments, yets have a call about it!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  private fun isPublicApplicationValid(
+    bookingRequestDto: BookingRequestDto?,
+    application: Application,
+    existingBooking: Visit?,
+  ) {
+    val prison = application.prison
+    val prisoner = prisonerService.getPrisoner(application.prisonerId) ?: throw ValidationException("prisoner not found")
+
+    checkPrison(prison.code, prisoner.prisonCode)
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt:59]: raised https://dsdmoj.atlassian.net/browse/VB-4086 to throw multiple validation issues at the same time. Currently being worked on.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    checkSlotCapacity(bookingRequestDto, application, existingBooking)
+  }
+
+  private fun isStaffApplicationValid(
+    bookingRequestDto: BookingRequestDto?,
+    application: Application,
+    existingBooking: Visit?,
+  ) {
+    // check capacity for slot
+    checkSlotCapacity(bookingRequestDto, application, existingBooking)
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt:85]: raised https://dsdmoj.atlassian.net/browse/VB-4085 and code changes done under separate branch.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:102]: to be included in https://dsdmoj.atlassian.net/browse/VB-4085

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/client/PrisonApiClient.kt [hunk 4]
```diff
@@ -42,6 +45,25 @@ class PrisonApiClient(
...
+
+  fun getVisitBalancesAsMono(prisonerId: String): Mono<Optional<VisitBalancesDto>> {
+    return webClient.get()
+      .uri("/api/bookings/offenderNo/$prisonerId/visit/balances")
+      .retrieve()
+      .bodyToMono<Optional<VisitBalancesDto>>()
+      .onErrorResume { e ->
+        if (e is WebClientResponseException && isNotFoundError(e)) {
+          // return an Optional.empty element if 404 is thrown
+          return@onErrorResume Mono.just(Optional.empty())
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/client/PrisonApiClient.kt]: remand prisoners will not have VOs and will return a 404. Handled as part of the updated getVisitBalances call and also added tests to check 404 responses.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  fun isApplicationValid(
+    bookingRequestDto: BookingRequestDto?,
+    application: Application,
+    existingBooking: Visit?,
+  ) {
+    try {
+      when (application.userType) {
+        UserType.PUBLIC -> isPublicApplicationValid(bookingRequestDto, application, existingBooking)
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/ApplicationValidationService.kt]: Updated

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**